<a href="https://colab.research.google.com/github/Rhabibi1609/Assignment/blob/main/YBIFoundation_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Diabetes Prediction**

---
### Objective:
###### *Predicting Diabetes using Random Forest Classification*


---

### Data Source:
https://www.kaggle.com/datasets/iammustafatz/diabetes-prediction-dataset/download?datasetVersionNumber=1

About the Dataset:
The Diabetes prediction dataset is a collection of medical and demographic data from patients, along with their diabetes status (positive or negative). The data includes features such as age, gender, body mass index (BMI), hypertension, heart disease, smoking history, HbA1c level, and blood glucose level. This dataset can be used to build machine learning models to predict diabetes in patients based on their medical history and demographic information. This can be useful for healthcare professionals in identifying patients who may be at risk of developing diabetes and in developing personalized treatment plans. Additionally, the dataset can be used by researchers to explore the relationships between various medical and demographic factors and the likelihood of developing diabetes.


---



# Import Library:

In [48]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import matplotlib.pyplot as plt
import seaborn as sns
import io
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

# Import Data:

In [1]:
from google.colab import files
 
 
uploaded = files.upload()

Saving diabetes_prediction_dataset.csv to diabetes_prediction_dataset.csv


In [6]:

df = pd.read_csv(io.BytesIO(uploaded['diabetes_prediction_dataset.csv']))


In [7]:
df.head()

gender   age  hypertension  heart_disease smoking_history    bmi  \
0  Female  80.0             0              1           never  25.19   
1  Female  54.0             0              0         No Info  27.32   
2    Male  28.0             0              0           never  27.32   
3  Female  36.0             0              0         current  23.45   
4    Male  76.0             1              1         current  20.14   

   HbA1c_level  blood_glucose_level  diabetes  
0          6.6                  140         0  
1          6.6                   80         0  
2          5.7                  158         0  
3          5.0                  155         0  
4          4.8                  155         0

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   gender               100000 non-null  object 
 1   age                  100000 non-null  float64
 2   hypertension         100000 non-null  int64  
 3   heart_disease        100000 non-null  int64  
 4   smoking_history      100000 non-null  object 
 5   bmi                  100000 non-null  float64
 6   HbA1c_level          100000 non-null  float64
 7   blood_glucose_level  100000 non-null  int64  
 8   diabetes             100000 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 6.9+ MB


# Describe Data:

In [10]:
df.describe()

age  hypertension  heart_disease            bmi  \
count  100000.000000  100000.00000  100000.000000  100000.000000   
mean       41.885856       0.07485       0.039420      27.320767   
std        22.516840       0.26315       0.194593       6.636783   
min         0.080000       0.00000       0.000000      10.010000   
25%        24.000000       0.00000       0.000000      23.630000   
50%        43.000000       0.00000       0.000000      27.320000   
75%        60.000000       0.00000       0.000000      29.580000   
max        80.000000       1.00000       1.000000      95.690000   

         HbA1c_level  blood_glucose_level       diabetes  
count  100000.000000        100000.000000  100000.000000  
mean        5.527507           138.058060       0.085000  
std         1.070672            40.708136       0.278883  
min         3.500000            80.000000       0.000000  
25%         4.800000           100.000000       0.000000  
50%         5.800000           140.000000       0.000000  
75%         6.200000           159.000000       0.000000  
max         9.000000           300.000000       1.000000

In [11]:
df.columns

Index(['gender', 'age', 'hypertension', 'heart_disease', 'smoking_history',
       'bmi', 'HbA1c_level', 'blood_glucose_level', 'diabetes'],
      dtype='object')

# Data Preprocessing:

## Define Target Variable (y) and Feature Variables (X):

In [13]:
y = df['diabetes']

In [17]:
X = df.drop(['diabetes'],axis=1)

In [35]:
X.replace({'smoking_history':{'No Info': 0, 'never': 1, 'former': 2,'not current':3,'current':4,'ever':5}},inplace=True)
X.replace({'gender':{'Male': 0,'Female':1,'Other':2}},inplace=True)

In [36]:
X.head()

gender   age  hypertension  heart_disease  smoking_history    bmi  \
0       1  80.0             0              1                1  25.19   
1       1  54.0             0              0                0  27.32   
2       0  28.0             0              0                1  27.32   
3       1  36.0             0              0                4  23.45   
4       0  76.0             1              1                4  20.14   

   HbA1c_level  blood_glucose_level  
0          6.6                  140  
1          6.6                   80  
2          5.7                  158  
3          5.0                  155  
4          4.8                  155

# Train Test Split:

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X,y, train_size=0.8, random_state=2529)

In [38]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((80000, 8), (20000, 8), (80000,), (20000,))

# Modeling:

In [39]:
model = RandomForestClassifier()

In [40]:
model.fit(X_train,y_train)

RandomForestClassifier()

In [42]:
y_pred = model.predict(X_test)

# Prediction:

In [43]:
y_pred 

array([0, 0, 0, ..., 0, 0, 0])

# Model Evaluation:

In [45]:
confusion_matrix(y_test,y_pred)

array([[18252,    69],
       [  508,  1171]])

In [46]:
accuracy_score(y_test,y_pred)

0.97115

In [47]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98     18321
           1       0.94      0.70      0.80      1679

    accuracy                           0.97     20000
   macro avg       0.96      0.85      0.89     20000
weighted avg       0.97      0.97      0.97     20000



#Explaination:
The Model uses RandomForestClassifier to predict the data and it has an accuracy score of 97%